### DISCUSSION TITLE
Adding Postal Message ID in case it not provided by SMTP client

### DISCUSSION THREAD

**Discussion by dragoangel** [postalserver/postal#1492] [Feature suggestions]
No MID is bad practice, and some SMTP clients not adding it to email. Most Mail delivery servers will add own MID in case client not set it to original email, but Postal don't doing it. Is it posible to add such function or add option to enable it per server?

**Comment by willpower232** [comment]
Would this only be a problem if the message was sent via a raw endpoint? I think Postal constructs most of the other messages itself so those would have them in?

**Comment by dragoangel** [comment]
It the case if smtp endpoint was been used and smtp client not provided MID header due wrong configuration, but actually this not rare unfortunately.
If rest api was been used then Postal will create own MID, yes.

**Comment by willpower232** [comment]
yes that makes sense, are you not adding the message ID in your application then? or are you using some software which doesn't completely follow the spec?

**Comment by dragoangel** [comment]
yep, some existing software doesn't follow the specs =( and doesn't create MID, but my existing mail system add missing MID, due this I can't move some of my existing webapps to Postal. Missing MID is really bad for delivery :(.

Also I tried check if missing MID will be added to RAW email, but doesn't have success with doing it as I always get error:
```
status time flags data                                                                                                        
------ ---- ----- ----                                                                                                        
error   0.0       @{code=UnauthenticatedFromAddress; message=The From address is not authorised to send mail from thiserver}
```
when `/api/v1/send/message` with same from user and API key works fine, maybe I missed something, but can't find what.
Powershell Script that I use to try send RAW email:
```
$Server = 'https://postal.example.com'
$APIKey = 'XXXX'

$Function = '/api/v1/send/raw'
$Method = 'Post'
$ContentType = 'application/json'

$From = 'sender@example.com'
$To = 'recipient@example.org'
$RFC2882 = @"
From: ${From}
To: ${To}
Subject: Test Email
Email Body $(Get-Date)
"@

$Data = [Convert]::ToBase64String([System.Text.Encoding]::Unicode.GetBytes($RFC2882))

$ToArray = @(${To})

$Headers = @{
    'X-Server-API-Key' = ${APIKey};
}

$Body = @{
    'mail_from' = ${From};
    'rcpt_to' = ${ToArray};
    'data' = ${Data};
}

Invoke-RestMethod -Uri ${Server}${Function} -Method ${Method} -Headers ${Headers} -Body (${Body}|ConvertTo-Json) -ContentType $ContentType
```

**Reply by zdeneksvarc** [reply]
My experience with Postal 2.1.1

The Raw API endpoint performs a sender address check on the `From:` entry in DATA. The envelope can contain any address in `MAIL FROM`, Postal does not handle this. Don't ask me why.

The problem with your script will be related to the fact that From: has a slightly different format than `MAIL FROM`:

Try formatting it any way you like:
- `From: "Mr. Postal" <no-reply@postal.fqdn>`
- `From: <no-reply@postal.fqdn>`

If you now correctly encode the DATA to Base64, there will be no problem.

I'll also note that the `Message-ID` should add the first SMTP element in the process, the MUA. If the MUA is not sending the `Message-ID`, then I would address the issue at the MUA. Yes, the big SMTP servers will correct the Message-ID if they don't get it or get it in the wrong format, but it's beyond their job.

**Comment by dragoangel** [comment]
Just more info, from Wiki:
Message-IDs, if present, are generated by the client program sending the email (mail user agent, or MUA) or by the first mail server (mail submission agent, or MSA).
This just to say that this is normal behavior to add missing MID. Thanks

**Comment by willpower232** [comment]
> From address is not authorised to send mail

This means that the domain name you're using doesn't match the ones configured for the credential so you should be able to resolve that.

The raw endpoint does even less to the message before sending it, expecting you to have created it properly so definitely won't add the message ID.

The regular message endpoint appears to create a message ID as it receives the parts of the message separately.

**Comment by dragoangel** [comment]
> This means that the domain name you're using doesn't match the ones configured for the credential

I was checked that. I send successfully over same API key by not-RAW method, but get error over RAW method. So no, this doesn't due domain

**Reply by zdeneksvarc** [reply]
See https://github.com/postalserver/postal/discussions/1492#discussioncomment-3743614

**Comment by zdeneksvarc** [comment]
From my point of view Postal (2.1.1) works quite correctly with `Message-ID`. Postal does not interfere with `Message-ID` if receives it from the MUA. If Postal itself acts as MUA (non-raw HTTP API), then Postal properly adds the `Message-ID` in the format: `<d90dabfa-01b5-4684-b0eb-0a849a03d456@rp.postal-fqdn>`

If the MUA is not sending the `Message-ID` or is sending it incorrectly, then the problem needs to be addressed at the MUA, not at Postal.
